FileManager/BackgroundFileSystem.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace FileManager
+{
+    class BackgroundFileSystem
+    {
+        public string RootDirectory { get; private set; }
+        public string SearchPattern { get; private set; }
+        public SearchOption SearchOption { get; private set; }
+
+        private FileSystemWatcher fileSystemWatcher { get; set; }
+        private BlockingCollection<FileSystemEventArgs> directoryChangesEvents { get; set; }
+        private Task backgroundScanner { get; set; }
+        private List<string> fsCache { get; set; }
+
+        public string FindFile(string regexPattern, RegexOptions options)
+        {
+            lock (fsCache)
+            {
+                return fsCache.FirstOrDefault(s => Regex.IsMatch(s, regexPattern, options));
+            }
+        }
+
+        public void Init(string rootDirectory, string searchPattern, SearchOption searchOptions)
+        {
+            RootDirectory = rootDirectory;
+            SearchPattern = searchPattern;
+            SearchOption = searchOptions;
+
+            //Calling CompleteAdding() will cause background scanner to terminate.
+            directoryChangesEvents?.CompleteAdding();
+            fsCache?.Clear();
+            directoryChangesEvents?.Dispose();
+            fileSystemWatcher?.Dispose();
+
+            fsCache = Directory.EnumerateFiles(RootDirectory, SearchPattern, SearchOption).ToList();
+
+            directoryChangesEvents = new BlockingCollection<FileSystemEventArgs>();
+            fileSystemWatcher = new FileSystemWatcher(RootDirectory);
+            fileSystemWatcher.Created += FileSystemWatcher_Changed;
+            fileSystemWatcher.Deleted += FileSystemWatcher_Changed;
+            fileSystemWatcher.Renamed += FileSystemWatcher_Changed;
+            fileSystemWatcher.Error += FileSystemWatcher_Error;
+            fileSystemWatcher.IncludeSubdirectories = true;
+            fileSystemWatcher.EnableRaisingEvents = true;
+
+            //Wait for background scanner to terminate before reinitializing.
+            backgroundScanner?.Wait();
+            backgroundScanner = new Task(BackgroundScanner);
+            backgroundScanner.Start();
+        }
+
+        private void FileSystemWatcher_Error(object sender, ErrorEventArgs e)
+        {
+            Init(RootDirectory, SearchPattern, SearchOption);
```

Comment:
This confuses me

Reply:
There's a possibility that FileSystemWatcher has an error while waiting for a change in the watched directory. If that happens, BackgroundFileSystem can't be sure it didn't miss a change, so we need to re-initialize the whole thing.

Reply:
Are you sure the old scanner can't survive and continue running?

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher.error?view=net-5.0

"Occurs when the instance of FileSystemWatcher is unable to continue monitoring changes or when the internal buffer overflows."